### PR TITLE
Reproducible Build Environment for Android CI using Nix

### LINKED ISSUE #622: Documenting the android docker build
We have docker containers for building the Android package but I am not sure that we have documentation on how to update and generate the docker containers.

The containers are stored in the github container registry at the MapLibre org level at https://github.com/orgs/maplibre/packages

We have:

* android-ndk-r21b
* android-base

Linux builder is probably for Qt, right @ntadej?

Let's document how we can update the docker images for the android builds...

### PR DESCRIPTION
**Pull request by louwers**
This PR will close #622 one it is done. I am opening a draft PR in the hope to get some early feedback.

Other related issues are #604 and #327 as well as getting the the `android-ci.yml` workflow to run through again.

The Docker container that the Android CI runs in needed to be updated / fixed. I am using the [Nix](https://nixos.org/) package manager and language to set up the build environment in a declarative and reproducible fashion. It uses [android-nixpkgs](https://github.com/tadfisher/android-nixpkgs) which is updated daily from Google's Android SDK repositories. This allows us to trivially keep the build environment up to date (see `README.md` inside). Another future possibility is setting up CI for different Android versions with relative ease.

### CODE REVIEW

**Comment by louwers** [comment]
The checks and the tests run through: https://github.com/louwers/maplibre-gl-native/actions/runs/3949312660/jobs/6760359736

File: .github/workflows/android-ci.yml [hunk 1]
```diff
@@ -3,37 +3,33 @@ name: android-ci
...
+    #   - "src/**"
+    #   - "test/**"
+    #   - "vendor/**"
+    #   - ".gitmodules"
+    #   - "!**/*.md"
 
 jobs:
   build:
-    runs-on: ubuntu-20.04
+    runs-on: ubuntu-latest
```
Comment (ntadej) [inline .github/workflows/android-ci.yml:28]: Even though that it is a container, we decided at some point not to use latest for reproducibility 😊

**Comment by louwers** [comment]
I will still add a workflow that builds and pushes the container to MapLibre's registry.

**Comment by ntadej** [comment]
I'm a bit hesitant having yet another package manager 😊

Did native GitHub actions show irreproducibilities?

**Comment by louwers** [comment]
> I'm a bit hesitant having yet another package manager blush
> 
> Did native GitHub actions show irreproducibilities?

Yes I can understand that, which is why I opened the Draft PR. We were using a Dockerfile to build the container, which is not reproducible or easy to update. For now we would only be using Nix for this very specific purpose.

We can also use a Dockerfile that uses sdkmanager to install the Android SDK, like: https://github.com/docker-android-sdk/android-31

**Comment by ntadej** [comment]
The question is rather, do we even need this docker image? I build my app for Android on a regular Ubuntu runner.

**Comment by louwers** [comment]
> The question is rather, do we even need this docker image? I build my app for Android on a regular Ubuntu runner.

I would argue that that setting up the Android SDK and all dependencies can be quite finicky and is not something you would want to do every time CI runs if it cannot be done in a way that is guaranteed to be reproducible. With Nix ensuring reprocibility the Docker container is not really needed, there is even a [GitHub Action](https://github.com/marketplace/actions/install-nix) for it. 

However, I think it may still be useful to have the CI run in a container so that people on platforms can run the CI environment locally.

**Comment by louwers** [comment]
I have hit a problem, which is that Javascript based GitHub workflow actions do not work with bare containers, including ones built by Nix: https://github.com/actions/checkout/issues/334

 -  There is a workaround shared: https://github.com/actions/checkout/issues/334#issuecomment-716068696

An alternative would be to use a Ubuntu based runner with Nix. However if we go that route the `/nix/store` would ideally need to be cached between builds and for that we would need to set up Cachix...

Another alternative we can use a build on a container such as https://github.com/thyrlian/AndroidSDK and add our specific dependencies.

Lastly we can forgo Nix or building an image altogether like @ntadej suggests.

**Comment by louwers** [comment]
I ended up pushing through with the Nix setup, but without docker container. Although a docker container can be built as well.

I need to keep the `$HOME` environment variable set, because robolectric needs to writes something there:

https://github.com/robolectric/robolectric/blob/cf24404ab7a0debddf0592886d34f5b9d9278e08/plugins/maven-dependency-resolver/src/main/java/org/robolectric/internal/dependency/MavenDependencyResolver.java#L124

**Comment by louwers** [comment]
It finally ran through https://github.com/louwers/maplibre-gl-native/actions/runs/3962802416/jobs/6789917460

**Comment by louwers** [comment]
Something that I was running into as well: Gradle Android Plugin was trying to download and install specific SDK versions and build tools. Which failed, because the SDK directory is not writable. To fix this the correct version (that is already installed) needed to be set in the Gradle files.

https://nixos.org/manual/nixpkgs/stable/#notes-on-improving-build.gradle-compatibility

**Comment by wipfli** [comment]
GitHub actually includes many relvant packages into their standard ubuntu images. See here for a list of versions: https://github.com/actions/runner-images/blob/main/images/linux/Ubuntu2204-Readme.md

Maybe we could just use the same versions as them and use directly their images...

**Comment by louwers** [comment]
@wipfli I wasn't aware of this!

That's pretty nice. This image will be availabe already on the GitHub runners so it is a few minutes quicker because it doesn't have to pull in the dependencies. We have a less control over the build environment, but we don't rely on experimental technology that is frankly unfamiliar to a lot of people. If we don't expect that we don't need the flexibility then it's probably the sensible choice. @thehoneymad I will leave the decision to you.

See #712 for the Android CI fix based on the GitHub Ubuntu image.

**Comment by louwers** [comment]
Found this on Hacker News just now: https://zero-to-nix.com/

**Review by thehoneymad — APPROVED** [review]

**Comment by thehoneymad** [comment]
Took a bit to understand Nix. I approved the high level approach I see. I need one more day to leave an in depth review since now I understand Nix a little better.

**Comment by maxammann** [comment]
This looks quite awesome! I was planning on adding Nix to maplibre-rs aswell. I'm just not sure yet to which extend.

Nix has some benefits over just relying on the Github Actions environment.

1. Using Nix has the advantage that the environment is identical in the CI and locally. 
2. Also Github Actions can change unexpectedly. For instance in the last update of the Runners they did not include up-to-date apt metadata. This can be updated with an `apt update` in the CI. Still this was kind of unexpected because one could assume that the current mirrors are compatible with the predefined apt metadata.
3. You eventually want to switch to another CI system. Defining dependencies explicitly has the advantage that switching to another CI is much simpler.

I'm not saying that Nix is the solution here. I'm just saying that something which setups the build environment is a huge benefit. Be it Docker or a VM or whatever.

**Comment by louwers** [comment]
I will keep this open for a bit longer before merging either this or #712. It is an interesting technology and I think there are clear benefits to having a declarative build environment that is easily updatable and extendable. Also when someone changes the `flake.nix` file in a PR the build environment is changed as well and it will be used for the CI run of that PR, no need to build and push a new Docker container. But there are trade-offs, as always.

I will add a [cachix action](https://nix.dev/tutorials/continuous-integration-github-actions#setting-up-github-actions) which caches the `/nix/store` between builds to see what that does to the build time.

**Comment by maxammann** [comment]
The main issue I see with nix is that it won't work for some dependencies like XCode, Clang/MSVC on windows.

That is what's holding me back in maplibre-rs

**Comment by ntadej** [comment]
Also we should not forget updating the fixed versions of SDKs. What annoys me the most when I start developing something if the library I use does not support the latest compiler/Xcode/NDK/SDK.

**Comment by maxammann** [comment]
> Also we should not forget updating the fixed versions of SDKs. What annoys me the most when I start developing something if the library I use does not support the latest compiler/Xcode/NDK/SDK.

There is no way to ensure that maplibre-gl-js / maplibre-rs supports the latest version of a compiler. So I think that is what Nix solves. It fixes all the versions so everyone is using the same software.

**Comment by louwers** [comment]
> Also we should not forget updating the fixed versions of SDKs. What annoys me the most when I start developing something if the library I use does not support the latest compiler/Xcode/NDK/SDK.

This will be a whole lot easier without having to build a new Docker container, and the packages for Nix being continiously updated.

> The main issue I see with nix is that it won't work for some dependencies like XCode, Clang/MSVC on windows.

Yes, for iOS Nix is not an option. However both Windows and Linux _could_ use Nix for the Android build environment, either directly or through a Docker image.

**Comment by wipfli** [comment]
I am happy to try out the nix android ci environment. The goal is just to document it well enough that somebody else can pick it up. I suggest @louwers you document how to use the android ci nix environment locally, and @ovivoda reads your documentation. Let not communicate directly just to simulate the situation in which you @louwers would not be present...

**Review by thehoneymad — APPROVED** [review]
This leaves me in a dubious place. My recommendation is still to maintain a development branch.

The benefits of using Nix here is reproducible builds. But the caveat is only the Nix parts are reproducible, as of now we are only using Nix for build environment replication. Nothing more.

So, technically we still can not use the Github Actions file directly to reproduce a local build. Is there any other way we can use our existing Github Actions files locally?

File: platform/android/ci/flake.nix [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+          };
+          contents = paths ++ [ pkgs.dockerTools.usrBinEnv ];
+        };
+      in
+      {
+				devShells = {
+				  default = pkgs.mkShell {
+            inherit env;
+            packages = paths;
+          };
```
[the remark this answers is not in the document]
Comment (louwers) [reply platform/android/ci/flake.nix:82]: Me pushing code from a computer where I don't have a formatter set up. 😅 

I will fix it.

File: platform/android/Makefile [hunk 3]
```diff
@@ -310,16 +310,6 @@ endif
 android-configuration: gradle/configuration.gradle
 	cat gradle/configuration.gradle
 
-# Updates Android's vendor submodules
-.PHONY: android-update-vendor
-android-update-vendor: gradle/configuration.gradle
-	$(MBGL_ANDROID_GRADLE) -Pmapbox.abis=none updateVendorSubmodules
```
[the remark this answers is not in the document]
Comment (louwers) [reply platform/android/Makefile:316]: This https://github.com/maplibre/maplibre-gl-native/blob/fcfc9a685d9688bb0fa688c96c9094a3c4686e3b/platform/android/gradle/gradle-update-vendor-modules.gradle

File: platform/android/ci/README.md [hunk 4]
```diff
@@ -0,0 +1,51 @@
+# MapLibre GL Native for Android - CI Environment
```
[the remark this answers is not in the document]
Comment (louwers) [reply platform/android/ci/README.md:1]: Yes, do you think it is better to move this to the Rust book?

[the remark this answers is not in the document]
Comment (thehoneymad) [reply platform/android/ci/README.md:1]: Yeah. I think you are touching a great part of MapLibre Core. We should have a consistent place to go and read about MapLibre things.

File: platform/android/Makefile (hunk 3 above)
[the remark this answers is not in the document]
Comment (thehoneymad) [reply platform/android/Makefile:316]: *Facepalm ensues*

**Comment by louwers** [comment]
Will merge the Nix setup in one form or another at some point later. :slightly_smiling_face:

**Pull request closed without merging**